exchanges/gemini/gemini_websocket.go

```diff
@@ -261,15 +322,62 @@ func (g *Gemini) wsHandleData(respRaw []byte, curr currency.Pair) error {
 	}
 	if _, ok := result["type"]; ok {
 		switch result["type"] {
-		case "subscription_ack":
-			var result WsSubscriptionAcknowledgementResponse
+		case "l2_updates":
+			var l2MarketData *wsL2MarketData
+			err := json.Unmarshal(respRaw, &l2MarketData)
+			if err != nil {
+				return err
+			}
+			g.wsProcessUpdate(l2MarketData)
+		case "trade":
+			var result wsTrade
 			err := json.Unmarshal(respRaw, &result)
 			if err != nil {
 				return err
 			}
+
+			tSide, err := order.StringToOrderSide(result.Side)
+			if err != nil {
+				g.Websocket.DataHandler <- order.ClassificationError{
+					Exchange: g.Name,
+					Err:      err,
+				}
+			}
+
+			enabledPairs, err := g.GetEnabledPairs(asset.Spot)
+			if err != nil {
+				return err
+			}
+
+			format, err := g.GetPairFormat(asset.Spot, true)
+			if err != nil {
+				return err
+			}
+
+			pair, err := currency.NewPairFromFormattedPairs(result.Symbol, enabledPairs, format)
+			if err != nil {
+				return err
+			}
+
+			tradeEvent := trade.Data{
+				Timestamp:    time.Unix(result.Timestamp/1000, 0),
+				CurrencyPair: pair,
+				AssetType:    asset.Spot,
+				Exchange:     g.Name,
+				Price:        result.Price,
+				Amount:       result.Quantity,
+				Side:         tSide,
+				TID:          strconv.FormatInt(result.EventID, 10),
+			}
+			if g.IsSaveTradeDataEnabled() {
```

Comment:
This check may as well be at the start of this case, we don't do anything with trades if its not enabled